feat: async requests and load tests

IgnatovFedor opened the pull request.

litinsky committed:
Merge pull request #1 from deepmipt/dev

Dev

litinsky committed:
Merge pull request #2 from deepmipt/dev

Dev

litinsky committed:
Merge pull request #3 from deepmipt/dev

Added str conversion of agent response

litinsky committed:
Merge pull request #4 from deepmipt/dev

Added -d/--download option for DeepPavlov components download

Commit:
Merge branch 'dev'

Commit:
Merge branch 'master' of https://github.com/deepmipt/convai_router_bot_poller

Commit:
Merge branch 'dev'

IgnatovFedor committed:
Removed model from repo and redirected request to stand

Added asynchronous requests

IgnatovFedor committed:
Added block to keep messages order

IgnatovFedor committed:
Added tests

IgnatovFedor committed:
Changed test structure

IgnatovFedor committed:
fix in tests

IgnatovFedor committed:
Test rewrited in class

IgnatovFedor committed:
Test cases creation parameterized

IgnatovFedor committed:
Deleted server file

IgnatovFedor committed:
Added log summary after tests

IgnatovFedor committed:
Tests parametrized

IgnatovFedor committed:
Responses now with aoihttp, logging is synchronous

IgnatovFedor committed:
fix: added check of response code, refactor in log parsing

IgnatovFedor committed:
Log parser refactored. Added test fail checking

IgnatovFedor committed:
Added parameter for multiple repeats of same test

IgnatovFedor committed:
Added model request timeout and replaced variance to standard deviation

IgnatovFedor committed:
namedtuple renamed

IgnatovFedor committed:
feat: added types in signatures

litinsky commented on poller.py:
```diff
@@ -155,15 +166,15 @@ def _process(self, payload: dict) -> None:
     args = parser.parse_args()
 
-    pipeline_config_path = find_config(args.config_path)
+    model_url = args.model_url
```
There should be possibility to set `model_url` both in command line args and poller config file

IgnatovFedor replied:
Done

litinsky commented on tests/poller_tests.py:
```diff
@@ -0,0 +1,136 @@
+        self._test_configs = tests_keeper.get_configs()
+
+        # Start poller
+        class Crutch:
```
What is it for?

IgnatovFedor replied:
Crutch class is for replacing argument parser in poller without rewriting poller

IgnatovFedor replied:
Crutch removed. Wrapper class now is testing directly

IgnatovFedor committed:
feat: added model_url parameter to poller config

litinsky commented on poller.py:
```diff
@@ -1,24 +1,17 @@
+from pathlib import Path
+from typing import Optional
 
 logging.disable(logging.DEBUG)
```
You should not disable logging level in code. Please parametrise logging level in config and set corresponding level, not disable unnecessary.

litinsky commented on tests/poller_tests.py:
```diff
@@ -0,0 +1,136 @@
+import json
+import logging
+import numpy as np
+import poller
```
Please feat import order according PEP8

IgnatovFedor committed:
fix: import according PEP8

IgnatovFedor committed:
Poller's logger config moved to config file

IgnatovFedor committed:
feat: in test poller.main() replaced Wrapper(). Changed config file

IgnatovFedor committed:
Some parameters from config moved to script to remove duplication

IgnatovFedor committed:
fix: removed unnecessary requirements

IgnatovFedor committed:
fix: added dateutil to requirements

IgnatovFedor committed:
fix: dateutil wrong name fixed

IgnatovFedor committed:
fix: changed requirements

IgnatovFedor committed:
fix: changed to work with 'payload' field instead 'text'

IgnatovFedor committed:
fix: bug fixed

When response contain " symbol router bot failed.
Example:
Hello Games called it an \"important first step into the world of synchronous co-op
This is was response of odqa en to "Hello" question
Also fixed bug with launching poller without model argument

IgnatovFedor committed:
Merge branch 'fix/text-to-payload' into stand

litinsky commented on poller.py:
```diff
@@ -73,50 +59,68 @@ def _process_input(self, input_q: dict) -> None:
-                    self._send_results(*resp)
+        tasks = (self.loop.create_task(self._process_chats_batch(chats_batch,
+                                                                 msg_id,
+                                                                 chat_ids)) for msg_id, chats_batch in enumerate(batched_chats))
```
Here and further naming: `msg_id` -> `layer_id`

litinsky commented on poller.py:
```diff
@@ -73,50 +59,68 @@ def _process_input(self, input_q: dict) -> None:
+        batch = list(zip(*chunk))
+        ids_batch = list(batch[0])
+        utts_batch = list(batch[1])
+        data = {"text1": utts_batch}
```
Parametrize: `"text1"` in `{"text1": utts_batch}`

litinsky commented on poller.py:
```diff
@@ -73,50 +59,68 @@ def _process_input(self, input_q: dict) -> None:
+            response = requests.Response()
+            response.status_code = 503
+        if response.status_code == 200:
+            tasks = (self.loop.create_task(self._send_results(*resp, msg_id)) for resp in zip(ids_batch, response.json()))
```
`self._send_results(*resp, msg_id)` - > ``self._send_results(resp[0], resp[1], msg_id)``

litinsky replied:
It is better to rewrite it in more explicit style.

litinsky commented on poller.py:
```diff
@@ -73,50 +59,68 @@ def _process_input(self, input_q: dict) -> None:
+                                                                 chat_ids)) for msg_id, chats_batch in enumerate(batched_chats))
+        await asyncio.gather(*tasks)
+
+    async def _process_chats_batch(self, chats_batch: list, msg_id: int, chat_ids: list) -> None:
```
Here, before and further use type annotation from `typing` if possible: for example `List[int]` or `Dict[str, List[str]]`.

IgnatovFedor replied:
Done for lists. Dicts left unchanged due to heterogeneous content.

litinsky commented on poller.py:
```diff
@@ -73,50 +59,68 @@ def _process_input(self, input_q: dict) -> None:
+        await asyncio.gather(*(self.loop.create_task(self._process_chunk(chunk, msg_id)) for chunk in chunked_utts_batch))
+
+    async def _process_chunk(self, chunk: list, msg_id: int) -> None:
+        batch = list(zip(*chunk))
```
Naming `batch` `->` `ids_utts_batch`.

litinsky commented on poller.py:
```diff
@@ -73,50 +59,68 @@ def _process_input(self, input_q: dict) -> None:
+        except requests.exceptions.ReadTimeout:
+            response = requests.Response()
+            response.status_code = 503
+        if response.status_code == 200:
```
```suggestion
        
        if response.status_code == 200:
```

IgnatovFedor replied:
Done

litinsky commented on poller.py:
```diff
@@ -73,50 +59,68 @@ def _process_input(self, input_q: dict) -> None:
+        else:
+            self.log.error(f'Got {response.status_code} code from {self.config["model_url"]}')
+            tasks = (self.loop.create_task(self._send_results(*resp, msg_id)) for resp in zip_longest(ids_batch, [], fillvalue='Server error'))
+        await asyncio.gather(*tasks)
```
```suggestion
        
        await asyncio.gather(*tasks)
```

IgnatovFedor replied:
Done

litinsky commented on poller.py:
```diff
@@ -73,50 +59,68 @@ def _process_input(self, input_q: dict) -> None:
+            tasks = (self.loop.create_task(self._send_results(*resp, msg_id)) for resp in zip(ids_batch, response.json()))
+        else:
+            self.log.error(f'Got {response.status_code} code from {self.config["model_url"]}')
+            tasks = (self.loop.create_task(self._send_results(*resp, msg_id)) for resp in zip_longest(ids_batch, [], fillvalue='Server error'))
```
May be tasks tuple should be composed once, after all checks.

IgnatovFedor committed:
fix: refactor

IgnatovFedor commented on poller.py:
```diff
@@ -73,50 +59,71 @@ def _process_input(self, input_q: dict) -> None:
+        return message
 
+    async def _send_results(self, chat_id: int, response: list, layer_id: int) -> None:
+        buf = {'text': ' '.join(str(element) for element in response)}
```
If model's response is `['']` (for example ODQA could send such response) then resp_text (see below) equals `{'text': ''}`. Router bot not sending any message to user with empty `text` field. What better to send in such cases? Maybe `-` to give user information that model respond nothing and not just ignored last message?

litinsky replied:
Not need for now, I think, we should simply translate model answers until router-bot user say to change this behavior.
